### PR TITLE
Derived Attribute: Bound Electron Density

### PR DESCRIPTION
**Pull request by ax3l**
Adds a new attribute to derive the density of bound electrons of an ion species.
 

## RT Test: FoilLCT at step 2000

ne
ne_bnd_c
ne_bnd_h
ne_bnd_n

### CODE REVIEW

File: include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.hpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+{
+namespace particleToGrid
+{
+namespace derivedAttributes
+{
+
+    HDINLINE float1_64
+    BoundElectronDensity::getUnit() const
+    {
+        const float_64 UNIT_VOLUME = ( UNIT_LENGTH * UNIT_LENGTH * UNIT_LENGTH );
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.hpp]: please remove brackets and use `constexpr`

```
constexpr float_64 UNIT_VOLUME = UNIT_LENGTH * UNIT_LENGTH * UNIT_LENGTH;
```

File: include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.hpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/particles/particleToGrid/derivedAttributes/ChargeDensity.def"
+
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.hpp]: optional: remove the empty line

File: include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.def [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        }
+
+        /** Calculate a new attribute  per particle
+         *
+         * Returns a new (on-the-fly calculated) attribute of a particle
+         * that can then be mapped to the cells the particle contributes to.
+         * This method is called on a per-thread basis (each thread of a block
+         * handles a particle of a frame).
+         *
+         * \tparam T_Particle particle in the frame
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.def]: please use `@` for doxygen comments

**Comment by ax3l** [comment]
> This probably does not need to be active by default for many cases as it can be retrieved from the ion density, right?

This can not be derived from the ion density for partly ionized ions :) I therefore added it to the FoilLCT example as a good showcase.

**Comment by n01r** [comment]
Oh :man_facepalming: of course not. When I did it I also always used the particle data.
You're right!

**Comment by ax3l** [comment]
One can derive it still as well from the ion charge density for a species with constant proton number, but sometimes one just wants the electron density of all physical electron reservoirs in a simple manner (e.g. for scattering).

**Comment by n01r** [comment]
Mmh, I happen to have done that before ^^

**Comment by ax3l** [comment]
@theZiz this is the other field we could visualize the FFT on (bound electrons and free electrons)

**Comment by ax3l** [comment]
@PrometheusPi feel free to review and merge

File: include/picongpu/simulation_defines/param/fileOutput.param [hunk 4]
```diff
@@ -39,6 +39,8 @@ namespace picongpu
      *
      * you can choose any of these particle to grid projections:
      *   - Density: particle position + shape on the grid
+     *   - BoundElectronDensity: density of bound electrons
+     *       note: only makes sense for partly ionized ions
```
Comment (PrometheusPi) [inline include/picongpu/simulation_defines/param/fileOutput.param]: REQUIRED: 
typo `partly` -> `partially`

File: share/picongpu/examples/FoilLCT/include/picongpu/simulation_defines/param/fileOutput.param [hunk 5]
```diff
@@ -39,6 +39,8 @@ namespace picongpu
      *
      * you can choose any of these particle to grid projections:
      *   - Density: particle position + shape on the grid
+     *   - BoundElectronDensity: density of bound electrons
+     *       note: only makes sense for partly ionized ions
```
Comment (PrometheusPi) [inline share/picongpu/examples/FoilLCT/include/picongpu/simulation_defines/param/fileOutput.param]: REQUIRED:
typo - see above

File: share/picongpu/examples/SingleParticleTest/include/picongpu/simulation_defines/param/fileOutput.param [hunk 6]
```diff
@@ -39,6 +39,8 @@ namespace picongpu
      *
      * you can choose any of these particle to grid projections:
      *   - Density: particle position + shape on the grid
+     *   - BoundElectronDensity: density of bound electrons
+     *       note: only makes sense for partly ionized ions
```
Comment (PrometheusPi) [inline share/picongpu/examples/SingleParticleTest/include/picongpu/simulation_defines/param/fileOutput.param]: REQUIRED:
typo - see above

File: include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.def [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+namespace picongpu
+{
+namespace particles
+{
+namespace particleToGrid
+{
+namespace derivedAttributes
+{
+    /** Density of Bound Electrons Operation for Particle to Grid Projections
```
Comment (PrometheusPi) [inline include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.def]: OPTIONAL:
This sentence sounds not correct - is there a better wording?

File: include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.def [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+     *
+     * Derives a scalar density field from a particle species at runtime.
+     * Each value is mapped per cell according to the species' spatial shape.
+     *
+     * @note only makes sense for party ionized ions
```
Comment (PrometheusPi) [inline include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.def]: REQUIRED: 
typo `party` -> `partially`

File: include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.def [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+           return unitDimension;
+        }
+
+        HINLINE std::string
+        getName() const
+        {
+            return "boundElectronDensity";
+        }
+
+        /** Calculate a new attribute  per particle
```
Comment (PrometheusPi) [inline include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.def]: OPTIONAL:
remove double space

**Comment by theZiz** [comment]
@ax3l: @HighIander already told me that you are working on a way to get the combined electron densities. Looking forward to it! 👍

File: include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.def (hunk 7 above)
Comment (ax3l) [reply include/picongpu/particles/particleToGrid/derivedAttributes/BoundElectronDensity.def]: that's how we document all of them - not gentle for long names but okish

**Comment by ax3l** [comment]
> @HighIander already told me that you are working on a way to get the combined electron densities. Looking forward to it! 👍

I thought ISAAC is able to combine two simple sources into one with a simple `add`, isn't it? ;)
But seriously: use the free e- density for the demo, the target is not of high Z so you won't see much difference (yet).

**Pull request merged**